article: tips for drift detection (#1586)
pulumi/pulumi-hugo: themes/default/content/blog/patterns-drift-detection/index.md

# A placeholder image representing the recommended format, dimensions and aspect
# ratio has been provided for you.
meta_image: meta.png

# At least one author is required. The values in this list correspond with the `id`
# properties of the team member files at /data/team/team. Create a file for yourself
# if you don't already have one.
authors:
    - david-flanagan

# At least one tag is required. Lowercase, hyphen-delimited is recommended.
tags:
    - continuous-delivery

# See the blogging docs at https://github.com/pulumi/pulumi-hugo/blob/master/BLOGGING.md.
# for additional details, and please remove these comments before submitting for review.
---

Let's face it, at some point someone is going to modify your carefully-crafted and automated infrastructure without updating your Pulumi program. These changes cause the desired state of our Pulumi program's to be inconsistent with the state of the world. These inconsistencies are often referred to as "drift". In this article, I want to cover a couple of patterns for detecting and reconciling this drift with your Pulumi programs.

<!--more-->

## Why is Drift Detection Important?

> Reconciling drift as quickly as possible is very important in the infrastructure as code (IaC) space because the longer drift is left unattended, the harder it becomes for automation to reconcile in a manner that can be considered safe for your organization, SLAs, and customers happiness. Imagine this as if it were your car. If a single tyre needs replaced, you might be OK for a while and not notice any problems driving. If you leave that unattended and another tyre runs flat, then you've got a slightly more serious problem. As these smaller problems accumulate you're applying a multiplier of risk to your overall infrastructure, especially during reconciliation. The more resources Pulumi needs to handle during this step, the more likely for some collateral damage.

Reviewer: ```suggestion
Reconciling drift as quickly as possible is very important in the infrastructure as code (IaC) space because the longer drift is left unattended, the harder it becomes for automation to reconcile in a manner that can be considered safe for your organization, SLAs, and customers' happiness. Imagine this as if it were your car. If a single tyre needs replaced, you might be okay for a while and not notice any problems driving. If you leave that unattended and another tyre runs flat, then you've got a slightly more serious problem. As these smaller problems accumulate, you're applying a multiplier of risk to your overall infrastructure, especially during reconciliation. The more resources Pulumi needs to handle during this step, the more likely for some collateral damage.
```

## Making Refresh Your Default

Before you go updating all your workflow files, or adopting the operator, for all the benefits of continuous reconciliation and drift detection, let's cover a few safety tips.

Always run a `pulumi up --refresh` locally / manually first and inspect the changes. Pulumi may wish to replace some resources and if you don't want that to happen - you need to prepare those resources for your new continuous reconciliation.

### Protect

If you have a resource that should NEVER be deleted, make sure you protect it. You can do this by adding a `ResourceOption` to the resource in your Pulumi program.

{{< chooser language "typescript,go,python,csharp,java,yaml" >}}

{{% choosable language typescript %}}

```typescript
{
    protect: true
}
```

{{% /choosable %}}

{{% choosable language python %}}

```python
opts=ResourceOptions(protect=True)
```

{{% /choosable %}}

{{% choosable language go %}}

```python
pulumi.Protect(true)
```

{{% /choosable %}}

{{% choosable language csharp %}}

```csharp
new CustomResourceOptions { Protect = true }
```

{{% /choosable %}}

{{% choosable language java %}}

```java

> CustomResourceOptions.builder()
>         .protect(true)

Reviewer: ```suggestion
CustomResourceOptions.builder()
        .protect(true)
        .build()
```

```

{{% /choosable %}}

{{% choosable language yaml %}}

```yaml
options:
  protect: true
```

{{% /choosable %}}

{{% /chooser %}}

### Ignore Changes

If Pulumi wants to replace a resource that you don't believe needs to be updated, you can ask Pulumi to `ignoreChanges` on certain properties. Of course, it might be best to have your code match those properties, but in the event that you can't - you can use the code below to ignore:

{{< chooser language "typescript,go,python,csharp,java,yaml" >}}

{{% choosable language typescript %}}

```typescript
{
    ignoreChanges: ["prop"]
}
```

{{% /choosable %}}

{{% choosable language python %}}

```python
opts=ResourceOptions(ignore_changes=["prop"])
```

{{% /choosable %}}

{{% choosable language go %}}

```python
pulumi.IgnoreChanges([]string{"prop"})
```

{{% /choosable %}}

{{% choosable language csharp %}}

```csharp
new CustomResourceOptions { IgnoreChanges = { "prop" } }
```

{{% /choosable %}}

{{% choosable language java %}}

```java

> CustomResourceOptions.builder()
>         .ignoreChanges("prop")

Reviewer: ```suggestion
CustomResourceOptions.builder()
        .ignoreChanges("prop")
        .build()
```

```

{{% /choosable %}}

{{% choosable language yaml %}}

```yaml
options:
    ignoreChanges:
    - prop
```

{{% /choosable %}}

{{% /chooser %}}

### Delete Before Replace

If you have any resources that don't use auto-naming, you'll also need to ensure they have the `deleteBeforeReplace` option enabled.

{{< chooser language "typescript,go,python,csharp,java,yaml" >}}

{{% choosable language typescript %}}

```typescript
{
    deleteBeforeReplace: true
}
```

{{% /choosable %}}

{{% choosable language python %}}

```python
opts=ResourceOptions(delete_before_replace=["prop"])
```

{{% /choosable %}}

{{% choosable language go %}}

```python
pulumi.DeleteBeforeReplace(true)
```

{{% /choosable %}}

{{% choosable language csharp %}}

```csharp
new CustomResourceOptions { DeleteBeforeReplace = true }
```

{{% /choosable %}}

{{% choosable language java %}}

```java

> CustomResourceOptions.builder()
>         .deleteBeforeReplace(true)

Reviewer: ```suggestion
CustomResourceOptions.builder()
        .deleteBeforeReplace(true)
        .build()
```
